src/components/atoms/Plugin/IFrame/hooks.ts

```diff
@@ -132,18 +132,34 @@ export default function useHook({
     doc.body.innerHTML = html;
 
     // exec scripts
-    Array.from(doc.body.querySelectorAll("script"))
-      .map<[HTMLScriptElement, HTMLScriptElement]>(oldScript => {
-        const newScript = document.createElement("script");
-        for (const attr of Array.from(oldScript.attributes)) {
-          newScript.setAttribute(attr.name, attr.value);
-        }
-        newScript.appendChild(document.createTextNode(oldScript.innerText));
-        return [oldScript, newScript];
-      })
-      .forEach(([oldScript, newScript]) => {
-        oldScript.parentNode?.replaceChild(newScript, oldScript);
-      });
+    const scripts = doc.body.querySelectorAll("script");
+
+    if (scripts) {
+      Array.from(scripts)
+        .filter(
+          script =>
+            script.getAttribute("type") !== "module" &&
+            !script.getAttribute("async") &&
+            !script.getAttribute("defer"),
+        )
+        .reduce((chain, oldScript) => {
+          return chain.then(() => runScript(oldScript));
+        }, Promise.resolve())
+        .finally(() => {
+          loaded.current = true;
+          onLoad?.();
+          Array.from(scripts)
+            .filter(
+              script =>
+                script.getAttribute("type") === "module" ||
+                script.getAttribute("async") ||
+                script.getAttribute("defer"),
+            )
+            .reduce((chain, oldScript) => {
+              return chain.then(() => runScript(oldScript));
+            }, Promise.resolve());
+        });
+    }
 
     // post pending messages
```

Comment:
Send the message after the scripts are loaded. We want to send it even if the script is not present, so we want to run it when loaded is true. You can cut this out to a separate function.

Reply:
Thank you! I've wrapped message sending part and loaded part into a seprate function, please help re-review it. Also i've update some on async scripts i think there's no need for load them one by one.